expr/func_movingwindow.go

```diff
@@ -0,0 +1,148 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/schema"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average", xFilesFactor: 0}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize, validator: []Validator{IsSignedIntervalString}},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncMovingWindow) Context(context Context) Context {
+	var err error
+	s.shiftOffset, err = s.getWindowSeconds()
+	if err != nil {
+		// shouldn't happen, validated above
```

Comment:
Should we log or something for visibility?

Reply:
I can do that as a fallback and log the error received.

For the case where where the user specified windowSize as points, the IsSignedIntervalString validator would catch it and proxy it around to graphite, so we wouldn't not see that message here (if that was your intention)

Reply:
No, that wasn't the intention. Just as a sanity check. You never know when someone might change `getWindowSeconds` in the future and make things very confusing.